ping/ping.md

```diff
@@ -0,0 +1,56 @@
+# Ping <!-- omit in toc -->
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+| --------------- | ------------------------ | ------ | --------------- |
+| 1A              | Candidate Recommendation | Active | r0, 2022-11-04  |
+
+Authors: [@marcopolo]
+
+Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]
+
+[@marcopolo]: https://github.com/mxinden
+[@mxinden]: https://github.com/mxinden
+[@marten-seemann]: https://github.com/marten-seemann
+
+# Table of Contents <!-- omit in toc -->
+- [Protocol](#protocol)
+- [Diagram](#diagram)
+
+# Protocol
+
+The ping protocol is a simple request response protocol. The client sends a
+payload of 32 random bytes, and the server responds with the same 32 bytes. The
+client then measures the RTT from the time it wrote the bytes to the time it
+received the bytes. The client may repeat the process by sending another payload
+with random bytes, so the server _should_ loop and echo the next payload. The
+client may close the write side of the stream after sending a payload, the
+server _should_ finish writing the echoed payload and then exit the loop and
+close the stream.
+
```

Comment:
```suggestion

The client MUST NOT open more than one outbound stream for the ping protocol.
```

How about making it explicit that the client only opens one stream at a time?

Unfortunately cross stream behavior is not linearizable for client and server. In other words, the client closing stream A and then opening stream B, might be perceived by the server as the client opening stream B and then closing stream A.

Still, I think the suggestion above allows the server to limit the amount of inbound streams to a small number, e.g. 2 and thus prevent DOS attacks where the client opens a large amount of ping streams.

What do folks think?

Reply:
Maybe this would be useful for testing multiple streams? e.g. I would like to measure the latency of N parallel streams doing a ping to smoke test my stream muxer implementation.

I agree on a bound, but maybe we can make the bound higher, like 64 streams.

Reply:
Thinking about this some more, I don’t think we need to support he testing use case here. For that we can have a separate testing protocol that can test things more precisely.

Agree on having this bound being 2 then.